framework/src/main/java/org/checkerframework/common/accumulation/AccumulationTransfer.java

```diff
@@ -0,0 +1,161 @@
+     * {@code a.b()}, and {@code a.b().c()} would have their estimates updated to include "foo". If
+     * any method in the chain is non-deterministic, its estimate will not be updated (but the rest
+     * of the chain is not affected).
+     *
+     * @param node a method invocation whose receiver is to be updated
+     * @param result the result containing the store to be modified
+     * @param values the new accumulation values
+     */
+    public void accumulate(
+            MethodInvocationNode node, TransferResult<CFValue, CFStore> result, String... values) {
+        Node receiver = node.getTarget().getReceiver();
+        AnnotatedTypeMirror oldType = typeFactory.getReceiverType(node.getTree());
+        // e.g. if the node being visited is static
+        if (oldType == null) {
+            return;
+        }
+        AnnotationMirror newAnno = getNewAnno(oldType, values);
+        while (receiver != null) {
+            // Note that this call doesn't do anything if receiver is a method call
+            // that is not deterministic, though it can still continue to recurse.
+            insertIntoStores(result, receiver, newAnno);
+
+            Tree receiverTree = receiver.getTree();
+            // Possibly recurse: if the receiver is itself a method call,
+            // then we need to also propagate this new information to its receiver
+            // if the method being called has an @This return type.
+            //
+            // Note that we must check for null, because the tree could be
+            // implicit (when calling an instance method on the class itself).
+            // In that case, do not attempt to refine either - the receiver is
+            // not a method invocation, anyway.
+            if (receiverTree == null || receiverTree.getKind() != Tree.Kind.METHOD_INVOCATION) {
+                // Do not continue, because the receiver isn't a method invocation itself. The
+                // end of the chain of calls has been reached.
+                break;
+            }
+
+            MethodInvocationTree receiverAsMethodInvocation =
+                    (MethodInvocationTree) receiver.getTree();
+
+            if (typeFactory.returnsThis(receiverAsMethodInvocation)) {
+                receiver = ((MethodInvocationNode) receiver).getTarget().getReceiver();
+            } else {
+                // Do not continue, because the method does not return @This.
+                break;
+            }
+        }
+    }
+
+    /**
+     * Combines the values in oldType with the values in newValues to produce a single accumulator
+     * type qualifier.
+     *
+     * @param oldType an annotated type mirror whose values should be included
+     * @param newValues new values to include
+     * @return an annotation representing all the values
+     */
+    private AnnotationMirror getNewAnno(AnnotatedTypeMirror oldType, String[] newValues) {
```

Comment:
`getNewAnno` is a vague name that I found confusing at call sites.  Please rename it.  Should "New" in the name be replaced by "Glb"?  (Or possibly by "Union"?)

Reply:
I prefer "combined", but if that's still confusing I can change it to "union", which I think sounds weird when you say it out loud